### ISSUE TITLE
bug: false positive in `no-derived-state` on state/props methods in subscription effect

### ISSUE THREAD

**Issue by nickjvandyke** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#36]
### Did you check docs and existing issues?

- [x] I have read the plugin docs
- [x] I have searched the existing issues
- [x] I have read [You Might Not Need An Effect](https://react.dev/learn/you-might-not-need-an-effect) and am reasonably confident that I *do* need an effect

### Plugin version

0.5.3

### ESLint version

9.0.0

### Describe the bug

Spinoff from https://github.com/NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect/issues/35

> Until now I have knowingly analyzed method calls on state as though they are state setters - I supposed mutating state is functionally equivalent to setting it. I would think that's generally correct, but I see now, not always. I will look into disabling that - it might make sense to trust the user knows what they're doing in this uncommon use.

### Steps To Reproduce

Lint the following code:

```js
        function Component() {
          const [name, setName] = useState();
          const [model] = useState(
            () => new FormModel(props)
          );

          // Register field within the model
          useEffect(() => {
            model.setFieldDescriptor(name);
            return () => model.removeField(name);
          }, [model, name]);
        }
```

### Expected Behavior

`model.setFieldDescriptor` is flagged, but should not be.

> However, model is kept in state as a singleton, and the subscription to it in an effect is legit, so I’m unsure why this would flag now
